py/malloc.c

```diff
@@ -263,6 +263,47 @@ void *m_tracked_calloc(size_t nmemb, size_t size) {
     return &node->data[0];
 }
 
+void *m_tracked_realloc(void *ptr_in, size_t n_bytes) {
+    // check for pure allocation
+    if (ptr_in == NULL) {
+        return m_tracked_alloc(n_bytes);
+    }
+    // check for pure free
+    if (n_bytes == 0) {
+        m_tracked_free(ptr_in);
+        return NULL;
+    }
+    m_tracked_node_t *node = (m_tracked_node_t *)((uint8_t *)ptr_in - sizeof(m_tracked_node_t));
+    size_t prev_bytes;
+    #if MICROPY_TRACKED_ALLOC_STORE_SIZE
+    prev_bytes = node->size;
+    #else
+    prev_bytes = gc_nbytes(node);
```

Comment:
Hi @dpgeorge I copied this logic from `free` below but feel there's a discrepancy here?  When `node->size` is set in alloc above it's set to `nmemb * size` so just the requested allocation size. 
I would have thought the `gc_nbytes(node)` version would return the entire allocated size, ie `sizeof(m_tracked_node_t) + nmemb * size` ?
Is this true, and if so does it not really matter as its only used for metrics reporting? I just found it a little confusing so would be happy to add some clarifying comments if needed, or update the impl to be consistent in both cases.

Reply:
FWIW I looked at this code last week, and my understanding is same for yours - this is only ever used for stats, either `MICROPY_TRACKED_ALLOC_STORE_SIZE` for debugging tracked alloc or complete memory stats when  `MICROPY_MALLOC_USES_ALLOCATED_SIZE` is set.

That said, maybe could we remove one of the config permutations by requiring `MICROPY_TRACKED_ALLOC_STORE_SIZE` to be set if `MICROPY_MALLOC_USES_ALLOCATED_SIZE` is set?

I also toyed with the idea of adding a `inline static size_t get_size(node)` function here to simplify some of these copy-paste ifdefs. I think I decided it was marginal for my PR, but if we're adding one more here it might be worth it!